### ISSUE TITLE
Failing builds with default parallelism run indefinitely

### ISSUE THREAD

**Issue by krlmlr** [ropensci/drake#427]
```r
library(drake)
plan <- drake_plan(
  a = stop(),
  b = a + 1
)

make(plan, jobs = 2, verbose = 3)
```

Run with `R -q -f ...` . Output:

```
...
> make(plan, jobs = 2, verbose = 3)
cache /home/kirill/git/R/drake/.drake
connect 1 import: plan
connect 2 targets: a, b
Error in sendMaster(try(lapply(X = S, FUN = FUN, ...), silent = TRUE)) : 
  write error, closing pipe to the master
Calls: make ... lightly_parallelize_atomic -> mclapply -> lapply -> FUN -> sendMaster
target a
fail a
Warning: Target `a` failed. Call `diagnose(a)` for details. Error message:
  
Call: 

```

It looks like the "mc" backend doesn't like it when the workers fail. I was able to change the behavior by demanding `keep_going = TRUE` or commenting out the `stop()` call in `drake_error()` . Maybe build error handling should always be done by the master in the case of manual scheduling?

**Comment by wlandau** [comment]
Good news: I can reproduce this one! We probably do not need that call to `stop()` inside `error_process()`. Failed workers already [log their errors and tracebacks in the "mc_fail" namespace](https://github.com/ropensci/drake/blob/2e65b13e473322f38c718eab491e54cdcfe6aa7a/R/handlers.R#L65), so the master could simply poll them and report back, gracefully killing the workers if `keep_going` is `FALSE`.

**Comment by wlandau** [comment]
Also, we should probably write to `mc_fail` only in `error_process()`, not `warning_process()`.

**Comment by wlandau** [comment]
Kirill, please try dee6e0905213a97f16c8ff60ead260f36af470aa. I think I eliminated the hanging. Errored workers log their errors in the cache's `"mc_error"` `storr` namespace, which the master process checks periodically.

**Comment by krlmlr** [comment]
Thanks, Will. I tried b102da12 which is a descendant of that commit (I believe). I'm seeing now:

```
...
target a
fail a
Warning message:
In mclapply(X = mc_worker_id(c(0, seq_len(config$jobs))), FUN = mc_process,  :
  scheduled cores 2, 1 encountered errors in user code, all values of the jobs will be affected
> 
```

Is this intended?

**Comment by wlandau** [comment]
Yes. The master process is one of those "scheduled cores", and it deliberately quits in error here. When it sees stuff in the `"mc_error"` `storr` namespace and `keep_going` is not explicitly `TRUE`, [it fails](https://github.com/ropensci/drake/blob/18cb853a42c44642301ae2304adc3e98296ecec6/R/mc_utils.R#L214) and [politely asks the workers to exit](https://github.com/ropensci/drake/blob/18cb853a42c44642301ae2304adc3e98296ecec6/R/mclapply.R#L58).

**Comment by wlandau** [comment]
Though I suppose it could just exit the while loop...

**Comment by wlandau** [comment]
Correction: what actually happens is not exactly what I intended after all. Currently, the worker fails if it's target fails (unless `keep_going` is `TRUE`) and the worker crash is passed on to the master. I think this is enough for this specific issue, but in general, I want to think about better ways to handle failures.

**Comment by wlandau** [comment]
Fixed in 5ca547d2a1d7c38b8a68639c74c4b15bd6b6d848. `mc_worker()` now exits more gracefully if it fails to build a target. Here is what I see now.

```r
library(drake)
clean(destroy = TRUE)
plan <- drake_plan(
  a = stop(123),
  b = a + 1
)
make(plan, jobs = 2)
#> target a
#> fail a
#> Error : Target `a` failed. Call `diagnose(a)` for details. Error message:
#>   123
```
